deepwell/src/methods/user.rs

```diff
@@ -0,0 +1,254 @@
+/*
+ * methods/user.rs
+ *
+ * DEEPWELL - Wikijump API provider and database manager
+ * Copyright (C) 2021 Wikijump Team
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+use self::users::Entity as User;
+use super::prelude::*;
+use crate::utils::replace_in_place;
+use sea_orm::DatabaseConnection;
+use wikidot_normalize::normalize;
+
+type Maybe<T> = Option<T>;
+
+#[derive(Deserialize, Debug)]
+struct CreateUser {
+    username: String,
+    email: String,
+    password: String,
+    language: Option<String>,
+}
+
+#[derive(Serialize, Debug)]
+struct CreateUserOutput {
+    id: i64,
+}
+
+pub async fn user_create(mut req: ApiRequest) -> ApiResponse {
+    let input: CreateUser = req.body_json().await?;
+    let db = &req.state().database;
+
+    // Derivative values
+    let slug = get_user_slug(&input.username);
+
+    // Check for conflicts
+    let result = User::find()
+        .filter(
+            Condition::any()
+                .add(users::Column::Username.eq(input.username.as_str()))
+                .add(users::Column::Email.eq(input.email.as_str()))
+                .add(users::Column::Slug.eq(slug.as_str())),
+        )
+        .one(db)
+        .await?;
+    if result.is_some() {
+        return error_response(StatusCode::Conflict, "");
+    }
+
+    // Insert new model
+    let user = users::ActiveModel {
+        username: Set(input.username),
+        slug: Set(slug),
+        email: Set(input.email),
+        email_verified_at: Set(None),
+        password: Set(input.password),
+        multi_factor_secret: Set(None),
+        multi_factor_recovery_codes: Set(None),
+        remember_token: Set(None),
+        language: Set(input.language),
+        karma_points: Set(0),
+        karma_level: Set(0),
+        real_name: Set(None),
+        pronouns: Set(None),
+        dob: Set(None),
+        bio: Set(None),
+        about_page: Set(None),
+        avatar_path: Set(None),
+        created_at: Set(Some(now())),
+        updated_at: Set(None),
+        deleted_at: Set(None),
+        ..Default::default()
+    };
```

Comment:
Just for the sake of my understanding: why is every field a `Set`?

Reply:
It is `Set` as opposed to `Unset`, it's a way of telling the active model that I'm only updating these fields, the rest should be left as-is.